### PR TITLE
Improve word2vec

### PR DESCRIPTION
**Pull request by unnonouno**
- Use multi dimensional embed ID
- Fix initial parameter
- Fix default minibatch size
- Merge kernels in negative sampling
- Fix default negative sampling size (=5)

With my GTX 970, its performance is about 300Kwords/sec with minbatch-size=5000, n_units=400, window=5, HSM and CBOW model. Negative sampling is slower than HSM because its implementation is not tuned.

### CODE REVIEW

File: examples/word2vec/train_word2vec.py [hunk 1]
```diff
@@ -58,11 +58,8 @@ def __init__(self, n_vocab, n_units, loss_func):
         )
 
     def __call__(self, x, context):
-        h = None
-        for c in context:
-            e = self.embed(c)
-            h = h + e if h is not None else e
-
+        e = model.embed(context)
+        h = F.sum(e, axis=0) * (1. / context.data.shape[0])
```
Comment (okuta) [inline examples/word2vec/train_word2vec.py]: How about `len(context.data)`?

File: examples/word2vec/train_word2vec.py [hunk 2]
```diff
@@ -90,23 +87,20 @@ def __init__(self, n_in, n_out):
         super(SoftmaxCrossEntropyLoss, self).__init__(
             W=L.Linear(n_in, n_out),
         )
+        self.W.W.data[...] = 0
```
Comment (okuta) [inline examples/word2vec/train_word2vec.py]: How about change the name `self.W`?

File: chainer/functions/loss/negative_sampling.py [hunk 3]
```diff
@@ -54,37 +54,37 @@ def forward_gpu(self, inputs):
...
         y = cuda.elementwise(
-            'T wx, int32 c, int32 m', 'T y',
+            'T wx, int32 n_samples', 'T y',
             '''
             T f = wx;
-            if (i % m == 0) {
+            if (i % n_samples == 0) {
               f = -f;
             }
             T loss;
```
Comment (okuta) [inline chainer/functions/loss/negative_sampling.py]: How about below code?
```python
T loss = 0;
if (f > 0) {
    loss = f;
    f = -f;
}
loss += log1pf(__expf(f));
```

File: examples/word2vec/train_word2vec.py [hunk 4]
```diff
@@ -194,7 +195,7 @@ def calculate_loss(model, dataset, position):
 
 with open('word2vec.model', 'w') as f:
     f.write('%d %d\n' % (len(index2word), args.unit))
-    w = model.embed.W.data
+    w = cuda.to_cpu(model.embed.W.data)
     for i in range(w.shape[0]):
```
Comment (okuta) [inline examples/word2vec/train_word2vec.py]: How about `for i, wi in enumerate(w):`?

**Comment by cemoody** [comment]
Hi @unnonouno, it might be possible to merge a lot of this PR with #1045. Both change the CUDA C code for NegativeSampling; one makes it faster and the other adds ignoring -1 labels.

**Comment by unnonouno** [comment]
Thank you very much! I'll review your PR #1045 first. And after that, I'll fix this PR.

**Comment by unnonouno** [comment]
I removed fixes related to the negative example as they are conflicted.

**Comment by okuta** [comment]
LGTM!
